keps/sig-node/20191025-image-pull-progress.md

```diff
@@ -0,0 +1,385 @@
+  necessary information from the underlying container runtime
+- Expose the progress to the Kubernetes API
+- Providing a way how to display the progress to the end user via the CLI
+
+### Non-Goals
+
+Everything which is not related to the container image pull procedure
+
+## Proposal / User Stories
+
+The overall implementation can be split up into three user stories.
+
+### CRI API enhancement
+
+The Kubernetes runtime API `ImageService` has to be modified to provide an
+additional streaming remote procedure call (RPC). This server-side stream
+provides continuously image pull related metadata during the whole image pull
+process. The new endpoint would be added in addition to the already available
+`PullImage` RPC and re-uses the request data type:
+
+```protobuf
+service ImageService {
+    …
+
+    // PullImageStream pulls an image and provides continuous metadata during
+    // that time
+    rpc PullImageProgress(PullImageRequest) returns (stream PullImageProgressResponse) {}
+}
+
+enum PullImageState {
+    STARTED = 0;
+    PULLING = 1;
+    FAILED  = 2;
+    DONE    = 3;
+}
+
+message PullImageProgressResponse {
+    // The current state of the image pull proress
+    PullImageState state = 1;
+
+    // The overall size of the image in bytes
+    uint64 size = 2;
+
+    // The amount of data already retrieved in bytes
+    uint64 current_offset = 3;
+
+    // String indicating the reason for the current state, like a failure message
+    string reason = 4;
+}
+```
+
+The `PullImageState` provides richer information to the kubelet in which state
+the image pull currently resides as well as a fine granular error messages
+directly from the runtimes via the `reason` field.
+
+### Kubernetes API additions
+
+The required API will be deployed via a Custom Resource Definition (CRD), which
+can be enabled via a feature gate. The definition applies cluster wide, whereas
+a single custom resource will be created on a per-node basis. This means that
+one custom resource gets managed per node, which contains the list of images and
+their current pull progress/status.
+
+The kubelet updates the Custom Resources during an image pull, which indicate
+the pull progress per container image. The amount of how often the kubelet
+updates the resources are configurable in a range between 1s and 1min.
+
+To reduce the impact of the added feature, a kubelet configuration option will
+be added which is (for now) disabled per default. This option completely avoids
+calling the new gRPC API and falls back to the current `PullImage` RPC.
+
+### CLI modifications
+
+The command line interface of Kubernetes (`kubectl`) does right now provide all
+necessary features to display the additional pull progress information.
+
+### Implementation Details/Notes/Constraints
+
+The major caveat for this implementation approach is the increased amount of
+resource updates from the kubelet during an image pull. This should be limited in a
```

Comment:
To preface, I'm far from a networking expert so I apologize if I'm misunderstanding. But at a high level, it feels a little strange that (in my opinion) image pull progress is most useful when the image is taking a long time to download (i.e. there's some form of network congestion), but the steps we are taking to show image pull progress could actually make the congestion worse, because the kubelet is doing more communication with the API server?

In other words, I'm not sure if its worth having the image pull take longer so that we can see its progress.